src/mono/wasi/runtime/main.c

```diff
@@ -10,7 +11,31 @@ const char* dotnet_wasi_getentrypointassemblyname();
 WASI_AFTER_RUNTIME_LOADED_DECLARATIONS
 #endif
 
-int main(int argc, char * argv[]) {
+#ifdef WASM_LIBRARY_MODE
+// _initialize is a function generated by the WASI SDK libc that calls the LLVM synthesized __wasm_call_ctors function for reactor components:
+// https://github.com/WebAssembly/wasi-libc/blob/9f51a7102085ec6a6ced5778f0864c9af9f50000/libc-bottom-half/crt/crt1-reactor.c#L7-L27
+// We define and call it for WASM_LIBRARY_MODE and TARGET_WASI to call all the global c++ static constructors.  This ensures the runtime is initialized
+// when calling into WebAssembly Component Model components.
+extern void _initialize();
+
+// CustomNativeMain programs are built using the same libbootstrapperdll as WASM_LIBRARY_MODE but wasi-libc will not provide an _initialize implementation,
+// so create a dummy one here and make it weak to allow wasi-libc to provide the real implementation for WASI reactor components.
+__attribute__((weak)) void _initialize()
+{
+}
+
+static bool runtime_initialized = false;
+
+#endif
+
+int initialize_runtime()
+{
+#if defined(WASM_LIBRARY_MODE)
+    if (runtime_initialized)
+		return 0;
+    _initialize();
+	runtime_initialized = true;
```

Comment:
This does not look thread safe.

(I know WASI is not MT ready in general, but it makes future work harder.)

Reply:
I agree, but currently, everything is single-threaded. We'll keep thread safety in mind for future work.